File: dsa/graph/dfs.js
const assert = require("assert")

const dfs = (graph, first) => {
    const stack = [ first ]
    const result = []
    while (stack.length > 0) {
        node = stack.pop()
        for (let i of graph[node]){
            stack.push(i)
        }
        result.push(node)
    }
    return result
}


const dfsRecursive = (graph, first) => {
    const recursive = (value) => {
        result.push(value)
        for (let v of graph[value]) {
            recursive(v)
        }
    }
    result = []
    recursive("a")
    return result
} 


test_dfs = () => {
    const graph = {
        "a": ["b", "c"],
        "b": ["d"],
        "c": ["e"],
        "d": ["f"],
        "e": [],
        "f": []
    }
    let data = dfsRecursive(graph, "a")
    let dataIterative = dfs(graph, "a")
    const output = ["a", "b", "d", "f", "c", "e"]
    const outputIterative = [ 'a', 'c', 'e', 'b', 'd', 'f' ]
    assert(JSON.stringify(data) == JSON.stringify(output))
    assert(JSON.stringify(dataIterative) == JSON.stringify(outputIterative))
}


test_dfs()
console.log("Test passert ==>")

File: dsa/graph/dfs.py
from typing import List

graph = {
    "a": ["b", "c"],
    "b": ["d"],
    "c": ["e"],
    "d": ["f"],
    "e": [],
    "f": []
}


def dfs(graph: dict) -> List[str]:
    """ traverse the graph in dfs order """
    keys = graph.keys()
    iterator = iter(keys)
    result = []
    stack = [ next(iterator) ]
    while stack:
        node = stack.pop()
        result.append(node)
        for i in graph[node]:
            stack.append(i)
    return result


def dfs_recursive(graph: dict) -> List[str]:
    """ traverse the graph in dfs order recursiverly """
    def recursive(value: str) -> List[str]:
        result.append(value)
        for i in graph[value]:
            recursive(i)

    keys = graph.keys()
    iterator = iter(keys)
    result = []
    recursive(next(iterator))
    return result


def test_dfs() -> None:
    """ test dfs function """
    result = dfs_recursive(graph)
    result2 = dfs(graph)
    print("recursive: ", result)
    print("iterative: ", result)
    assert (
        (result == ["a", "c", "e", "b", "d", "f"]
        or result == ["a", "b", "d", "f", "c", "e"])
        and
        (result2 == ["a", "c", "e", "b", "d", "f"]
        or result2 == ["a", "b", "d", "f", "c", "e"])
    )

if __name__ == "__main__":
    test_dfs()
    print(10*"=", " Passed tests ", 10*"=")
